### PR TITLE
uorb autogeneration

### PR DESCRIPTION
**Pull request by dagar**
I've split this out of the new logger work.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Neat! @julianoes Can you build this branch on Snapdragon to make sure it still works?

**Comment by julianoes** [comment]
@LorenzMeier I tested this on the Snapdragon and the ADSP crashes on ekf2 start.

**Comment by kd0aij** [comment]
on fmu-v4, I'm getting hardfaults with this branch. According to main.map, the fault occurs while in DriverFramework::DevHandle::~DevHandle()
```
Assertion failed at file:armv7-m/up_hardfault.c line: 184 task: commander
sp:     20003980
IRQ stack:
  base: 200039cc
  size: 000002e8
20003980: 080f841c 000000b8 080b49cd 00000010 200036cc 00000003 00000000 080b9159
200039a0: 080b9145 080c4753 e000ed24 080c2321 00000010 20001831 00000000 00000000
200039c0: 00000000 080b90c9 2000aa6c 00000000 00000000 10000010 00000001 00000000
sp:     2000ab40
User stack:
  base: 2000b4a8
  size: 00000e0c
2000ab40: 00000001 20001821 20001778 0803a06f 00000001 20001781 2000ac68 00000000
2000ab60: 2000ac6c 2000ac70 2000acd8 ffffffff a7a5c50f 00000000 00000000 00016682
2000ab80: 00000000 0000000a 00000000 00000001 00000000 00000012 a7a5c50f 00000000
2000aba0: 20009b40 00000005 00000008 00000009 0000000b 0000000d 0000000e 0000000f
2000abc0: 00000010 00000011 00000012 00000013 00000014 2000ad48 00000000 00000000
2000abe0: 00000105 000002ab 00000038 00000006 00000007 0000000c 00000015 00000016
2000ac00: 00000017 00000018 00000102 00000103 0000010c 0000010d 0000002f 00000034
2000ac20: 00000030 00000031 00000032 00000033 00000037 00000035 00000036 000000ce
2000ac40: 00000039 0000003a 000002ff 00000300 00000301 00000302 00000303 00000304
2000ac60: 1000f780 1000e920 00760000 2000fc90 2000f3a0 01010100 00000000 00000000
2000ac80: 00000000 0000000a 3f000000 00000000 00000001 3f000000 40a00000 41200000
2000aca0: 00000001 00000000 00000000 00000032 19dd9b49 00000001 00000000 00000000
2000acc0: fe71168e 3b091a2f ffff0001 005b55d6 00000000 ffff0001 00000000 00000000
2000ace0: ff000190 0000001a ffffffff ffffffff ffffffff 19de574f 00000001 408e00cd
2000ad00: 00010001 ffffff00 00000000 00000000 00000000 00000000 ff000000 ffffffff
2000ad20: ffffffff ffffffff ffffffff ffffffff 19cf98f4 00000001 00000000 00000000
2000ad40: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000ad60: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000ad80: 00000000 00000000 00000000 00000000 00000000 00000000 00000001 00000000
2000ada0: 00000000 00000000 00000000 00000001 00000000 19dd9851 00000001 19dd9624
2000adc0: 00000001 3b9441b5 bc05e35c 3ae36e2b 00000000 00000000 00000000 00000000
2000ade0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000ae00: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffff00
2000ae20: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000ae40: 00000000 00000000 00000000 00000190 00000001 00000000 000000ff 00000000
2000ae60: ffffff00 19de6294 00000001 a175bc2c 00053163 8de74cd0 4043e000 e626c0de
2000ae80: c05a4647 449c1ead 3dbb1e86 bda7c85d 3d032be3 bd207c10 409fef9e 40e4ac09
2000aea0: 00000000 44e1f80e ffff0000 19de6294 00000001 a175bc2c 00053163 8de74cd0
2000aec0: 4043e000 e626c0de c05a4647 449c1ead 3dbb1e86 bda7c85d 3d032be3 bd207c10
2000aee0: 409fef9e 40e4ac09 00000000 44e1f80e ffff0000 19de6294 00000001 045c580f
2000af00: 00000000 1309c800 4043e046 c76229c0 c05a4635 00000000 00000000 c36be8a5
2000af20: c2bd19f1 c04f94b7 3dbb1e86 bda7c85d 3d032be3 bd207c10 449bb6e3 00000000
2000af40: 00000000 409fef9e 40e4ac09 01010101 ff000101 00000000 00000000 19dcc654
2000af60: 00000001 19dcc663 00000001 19dcc653 00000001 19dcc653 00000001 a175bc2c
2000af80: 00053163 17b15f24 c15b4bee 001a0729 00000000 3f8978d5 3f32714d 409fef9e
2000afa0: 40e4ac09 3f9c28f6 3fce147b 00000067 00000014 3dae147b bc03126f bdae147b
2000afc0: bdfdf3b7 4090b359 ff010903 19de6126 00000001 bb4074b2 3b7b75f9 bd207c10
2000afe0: b8d06298 3a9c160c 397a6f20 00000000 00000000 00000000 3c688517 3ca1991f
2000b000: 3c85f76d b99fd676 bb51d5a4 ba02f963 3f7fcd3b 3d206592 3b816703 bd207140
2000b020: 3f7fcd76 3b367541 bb7b75d1 bb407443 3f7fff3e 3f7ff33e babb7e1a 3aff2ea3
2000b040: bca07368 3b7e2e10 3dbf860d c11c5eb0 ffff0101 00000000 00000000 00000000
2000b060: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b080: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b0a0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b0c0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b0e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b100: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b120: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b140: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b160: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b180: ff000000 19de6126 00000001 19de6126 00000001 19de5470 00000001 00000000
2000b1a0: 00000000 00000fa0 00000000 00000fa0 00000000 00000000 00000000 00000fa0
2000b1c0: 00000000 00000fa0 00000000 00000000 00000000 19de6126 00000001 19de5470
2000b1e0: 00000001 00000000 00000000 19de402c 00000001 19de2129 00000001 19de41ad
2000b200: 00000001 19de5b4c 00000001 00000000 00000000 00000000 00000000 00000000
2000b220: 00000000 00000000 00000000 00000000 00000000 3a4d3ebc 3b802e96 3a894fab
2000b240: bac34db1 b985ca0d bb29388e 00000000 00000000 00000000 00000064 00000063
2000b260: 00000000 355ca78d 379364b1 36463a8a b73259a7 b5f8ab3f b76ba641 00000000
2000b280: 00000000 00000000 00000000 00000000 00000000 42561cd4 4277adbd 00000000
2000b2a0: 3d36d93c 3cc4d9f7 c11c8ad3 bd21bd34 be29d870 c11d3a4f 00000000 00000000
2000b2c0: 00000000 394227d9 38a5c5f8 bd20e8ce b92c710f ba31cd5c bd2098af 00000000
2000b2e0: 00000000 00000000 00000000 00000000 00000000 00000064 00000063 00000000
2000b300: 00000000 00000000 00000000 42561cd4 4277adbd 00000000 3e5e7d87 3d1d77ec
2000b320: 3ebac411 3e5cb9ef 3e07d80c 3eded195 3e58e77f 3d3c796a 3ed300a8 00000000
2000b340: 00000000 00000000 00000000 00000000 00000000 000000ff 00000064 00000032
2000b360: 00000000 00000000 00000000 42180000 42630000 42779f8e 444b870a 00000000
2000b380: 00000000 44e1f80e 00000000 00000000 4239b852 00000000 00000000 0000004b
2000b3a0: 00000000 00000000 00000000 00000000 00000000 3d460000 3bed9999 40092800
2000b3c0: 3fd68000 3fa8eb33 3f8c2599 3f645199 3f368800 3ec2b333 00000000 00000000
2000b3e0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b400: 00000000 00000000 00000000 00000000 003afffe 0006ffec 00080012 00000000
2000b420: fe570000 ef080010 000cff8a 0000f814 00000000 00000000 00370000 021a004c
2000b440: fe4bfdf6 00ff0155 0227ff33 00000000 00020000 00040003 000b000a 000d000c
2000b460: 0010000e 00000000 00000000 00000000 00000000 00000000 00000000 00000000
2000b480: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 080b5ead
2000b4a0: 00000000 00000000 ffffffff c9d379fe 2ad5271c a18f8b1d 00000040 80000e20
R0: 19de65d9 00000001 19de0000 00000001 20001831 00000000 00000000 00000000
R8: 00000000 00000000 20001749 00000000 2000aaf0 2000ab40 080cf463 080cf464
xPSR: 01000000 BASEPRI: 00000010 CONTROL: 00000000
EXC_RETURN: ffffffe9
```

**Comment by julianoes** [comment]
> on fmu-v4, I'm getting hardfaults with this branch. According to main.map, the fault occurs while in DriverFramework::DevHandle::~DevHandle()

Aha that is an interesting pointer, thanks! I'll try to have a look sometime this week.

**Comment by kd0aij** [comment]
Does it make any sense that DriverFramework is involved?

**Comment by dagar** [comment]
The `-Wl,--gc-sections` change I mentioned in #4365 actually went in here.
If I had to guess...

**Comment by NaterGator** [comment]
Enable --print-gc-sections and see what is getting culled?

**Comment by dagar** [comment]
Yep

Removing unused section

    vtable for DriverFramework::DFUIntList
    vtable for DriverFramework::DFPointerList
    DriverFramework::DFUIntList::DFUIntListNode::DFUIntListNode(unsigned int)
    DriverFramework::DFUIntList::DFUIntListNode::~DFUIntListNode()
    DriverFramework::DFUIntList::get(DriverFramework::DFUIntList::DFUIntListNode*, unsigned int&)
    DriverFramework::DFUIntList::next(DriverFramework::DFUIntList::DFUIntListNode*&)
    DriverFramework::DFUIntList::size()
    DriverFramework::DFUIntList::clear()
    DriverFramework::DFUIntList::empty()
    DriverFramework::DFUIntList::erase(DriverFramework::DFUIntList::DFUIntListNode*)
    DriverFramework::DFUIntList::pushBack(unsigned int)
    DriverFramework::DFUIntList::pushFront(unsigned int)
    DriverFramework::DFUIntList::DFUIntList()
    DriverFramework::DFUIntList::~DFUIntList()
    DriverFramework::DFUIntList::~DFUIntList()
    DriverFramework::DisableCopy::~DisableCopy()
    DriverFramework::DisableCopy::~DisableCopy()
    DriverFramework::DFPointerList::DFListNode::DFListNode(void*)
    DriverFramework::DFPointerList::DFListNode::~DFListNode()
    DriverFramework::DFPointerList::get(DriverFramework::DFPointerList::DFListNode*)
    DriverFramework::DFPointerList::next(DriverFramework::DFPointerList::DFListNode*&)
    DriverFramework::DFPointerList::size()
    DriverFramework::DFPointerList::clear()
    DriverFramework::DFPointerList::empty()
    DriverFramework::DFPointerList::erase(DriverFramework::DFPointerList::DFListNode*)
    DriverFramework::DFPointerList::pushBack(void*)
    DriverFramework::DFPointerList::pushFront(void*)
    DriverFramework::DFPointerList::DFPointerList()
    DriverFramework::DFPointerList::~DFPointerList()
    DriverFramework::DFPointerList::~DFPointerList()
    DriverFramework::clockGetRealtime(timespec*)
    DriverFramework::clockGetMonotonic(timespec*)
    DriverFramework::absoluteTimeInFuture(unsigned long long)
    DriverFramework::DevMgr::initialize()
    DriverFramework::DevMgr::waitForUpdate(DriverFramework::DFPointerList&, DriverFramework::DFPointerList&, unsigned int)
    DriverFramework::DevMgr::getNextDeviceName(unsigned int&, char const**)
    DriverFramework::DevMgr::setDevHandleError(DriverFramework::DevHandle&, int)
    DriverFramework::DevMgr::finalize()
    DriverFramework::SyncObj::waitOnSignal(unsigned long)
    DriverFramework::SyncObj::lock()
    DriverFramework::SyncObj::signal()
    DriverFramework::SyncObj::unlock()
    DriverFramework::SyncObj::SyncObj()
    DriverFramework::SyncObj::~SyncObj()
    DriverFramework::DevHandle::write(void const*, unsigned int)
    DriverFramework::HandleObj::~HandleObj()
    DriverFramework::HandleObj::~HandleObj()
    DriverFramework::initMutex(pthread_mutex_s&)
    Sensors::apply_mag_calibration(DriverFramework::DevHandle&, mag_calibration_s const*, int)
    Sensors::apply_gyro_calibration(DriverFramework::DevHandle&, gyro_calibration_s const*, int)
    Sensors::apply_accel_calibration(DriverFramework::DevHandle&, accel_calibration_s const*, int)


We either need to do something to force it to be kept, or ifdef DriverFramework usage everywhere.

**Comment by kd0aij** [comment]
I don't suppose this explains the earlier hardfaults on the newLogger branch.

**Comment by dagar** [comment]
With the cmake `-Wl,--gc-sections` change on master I'm not seeing any hard faults and the same DriverFramework sections removed. Let's test and merge that separately. #4373

**Comment by julianoes** [comment]
I'll re-test tomorrow.

**Comment by mcharleb** [comment]
@julianoes Is there anything I should look at here for DriverFramework?

**Comment by kd0aij** [comment]
Ran the same pixracer for 2 hours and 15 minutes on this branch without a crash. No changes to the source, but did checkout and build/flash other branches before rebuilding and flashing this one.
```
2000_01_01_00_01_40: [cmd] ARMED by command line
2000_01_01_00_01_41: [blackbox] /fs/microsd/log/sess008
2000_01_01_00_01_41: [blackbox] recording: log001.px4log
2016_04_26_18_34_59: [cmd] DISARMED by arm/disarm component command
2016_04_26_18_35_00: [blackbox] stopped (14889 drops)
2016_04_26_19_23_57: [cmd] ARMED by arm/disarm component command
2016_04_26_19_23_57: [blackbox] /fs/microsd/log/2016-04-26
2016_04_26_19_23_57: [blackbox] recording: 19_23_57.px4log
2016_04_26_19_25_22: [cmd] DISARMED by arm/disarm component command
2016_04_26_19_25_23: [blackbox] stopped (568 drops)
2016_04_26_19_50_46: [inav] GPS signal found
2016_04_26_19_50_47: [inav] init ref: 39.7502382, -105.0979762, 1718.8
2016_04_26_19_52_34: [inav] GPS signal lost
2016_04_26_19_57_01: [inav] GPS signal found
2016_04_26_19_57_37: [inav] GPS signal lost
2016_04_26_20_00_57: [inav] GPS signal found
2016_04_26_20_02_42: [inav] GPS signal lost
2016_04_26_20_09_11: [inav] GPS signal found
2016_04_26_20_13_19: [inav] GPS signal lost
2016_04_26_20_15_29: [inav] GPS signal found
2016_04_26_20_21_58: [inav] GPS signal lost
2016_04_26_20_22_34: [inav] GPS signal found
2016_04_26_20_30_51: [cmd] ARMED by command line
2016_04_26_20_30_51: [blackbox] /fs/microsd/log/2016-04-26
2016_04_26_20_30_51: [blackbox] recording: 20_30_51.px4log
2016_04_26_20_40_00: TAKEOFF DETECTED
2016_04_26_20_40_04: LANDING DETECTED
2016_04_26_20_42_19: TAKEOFF DETECTED
2016_04_26_20_42_22: LANDING DETECTED
2016_04_26_20_43_03: TAKEOFF DETECTED
2016_04_26_20_43_06: LANDING DETECTED
2016_04_26_20_43_52: TAKEOFF DETECTED
2016_04_26_20_43_54: LANDING DETECTED
2016_04_26_20_44_19: [inav] GPS signal lost
2016_04_26_20_46_28: [inav] GPS signal found
2016_04_26_20_48_01: [inav] GPS signal lost
2016_04_26_20_48_30: [inav] GPS signal found
2016_04_26_20_49_32: TAKEOFF DETECTED
2016_04_26_20_49_35: LANDING DETECTED
2016_04_26_20_59_03: [inav] GPS signal lost
2016_04_26_20_59_04: [inav] GPS signal found
2016_04_26_20_59_07: [inav] GPS signal lost
2016_04_26_20_59_37: [inav] GPS signal found
```

**Comment by dagar** [comment]
@kd0aij the uorb header `#pragma pack(push, 1)` is causing each struct to be misaligned.
For example in fw_pos_ctrl the sensor_combined struct was at 0x20021337

I'm trying to use `__attribute__((aligned(4), packed))` instead, but there are other problems that GCC now catches like not being able to pass fields of a packed struct as a reference.

**Comment by kd0aij** [comment]
Interesting, but wouldn't unaligned accesses either fail completely, or always work?

**Comment by dagar** [comment]
For me it was always failing in the same place.

**Comment by kd0aij** [comment]
Every one of my hardfaults shows a return address on the stack at line 2295
of commander.cpp, but I don't see any connection between that line and orb
topics.

**Comment by dagar** [comment]
What's 2295 for you? In this branch and newLogger2 it's a comment block.
https://github.com/PX4/Firmware/blob/newLogger2/src/modules/commander/commander.cpp#L2295

**Comment by davids5** [comment]
@kd0aij @dagar 

Hmm - this might be stack related - you can enable the CONFIG_ARMV7M_STACKCHECK and set a break in the point it vectors to the hard fault from the check preamble.

Nuttx and the PX4 code need to be in agreement on this setting: So - If you uses it do a clean build, after setting or unsetting it.

**Comment by kd0aij** [comment]
I was behind a few commits on uorb_autogen. Just committed a workaround which solves my crash: is yours in a different place?

**Comment by dagar** [comment]
@kd0aij _home is a generated orb struct (home_position_s). Going back a few changes so you can reproduce your hard fault, what is `&_home.timestamp`? Your work around avoids the problem by not passing something unaligned to hrt_elapsed_time().

Thanks for the input @davids5. I'll give that a try tomorrow.

**Comment by kd0aij** [comment]
oh, I didn't realize you had already eliminated the problem by aligning the structs. I reverted my change and no problem so far. Is there any impact on logger efficiency with the new alighment?

**Comment by kd0aij** [comment]
and my workaround should go away... I assume you'll prefer to drop it during a rebase.

File: src/drivers/device/integrator.h [hunk 1]
```diff
@@ -69,7 +69,7 @@ class Integrator
 	 * @param integral_dt	Get the dt in us of the current integration (only if reset).
 	 * @return		the integral since the last read-reset
 	 */
-	math::Vector<3>		get(bool reset, uint64_t &integral_dt);
+	math::Vector<3>		get(bool reset, uint64_t *integral_dt);
```
Comment (julianoes) [inline src/drivers/device/integrator.h]: May I ask why this change? Any particular reason?
 
It broke `make eagle_default`.

**Comment by julianoes** [comment]
Just tested this again, and it still segfaults on the DSP side on Snapdragon.

File: src/drivers/device/integrator.h (hunk 1 above)
Comment (julianoes) [inline src/drivers/device/integrator.h]: I'll just push a commit to fix this.

**Comment by julianoes** [comment]
It crashes somewhere in commander, I'm doing some debugging.

**Comment by julianoes** [comment]
I narrowed the problem down but I can't figure out why exactly the crash happens.

The failing line is:
https://github.com/PX4/Firmware/blob/ec4d2fa8de963e1d9379aa32e57b1723619b7b0d/src/modules/commander/commander.cpp#L1740

The pointer and value at `&sensors.barometer_failure` are correct.

The crash is not happening when `hrt_elapsed_time()` is used without the `volatile` keyword:
The declaration is here: https://github.com/PX4/Firmware/blob/ec4d2fa8de963e1d9379aa32e57b1723619b7b0d/src/drivers/drv_hrt.h#L103

Alternatively, you can work around the bug like below but it won't prevent the same thing from happening anywhere else:

```
const hrt_abstime temp_timestamp = sensors.baro_timestamp[0];
if (hrt_elapsed_time(&temp_timestamp) < FAILSAFE_DEFAULT_TIMEOUT) {
```

@mcharleb, @LorenzMeier: any idea why Snapdragon would crash when calling the function which uses  has a `volatile` argument?

**Comment by dagar** [comment]
@julianoes you said &sensors.barometer_failure was correct, but was it 4 byte aligned? What compiler do you use for eagle? I'm guessing not gcc or clang? 

I borrowed this from mavlink to handle packing and alignment, but so far ignored the !GNUC part.

    // Macro to define packed structures
    #ifdef __GNUC__
    #define ORBPACKED( __Declaration__ ) __Declaration__ __attribute__((aligned(4), packed))
    #else
    #define ORBPACKED( __Declaration__ ) __pragma( pack(push, 1) ) __Declaration__ __pragma( pack(pop) )
    #endif


In general I don't like that we're messing with the packing and alignment of uorb structs. This is mainly for logger simplicity and performance so it can do a single copy per message. I'm going to see if we can avoid messing with packing and alignment by sorting the struct by size (already done) and adding some small filler where necessary.

File: src/drivers/device/integrator.h (hunk 1 above)
Comment (dagar) [inline src/drivers/device/integrator.h]: We really need to get every target we care about building in travis-ci or
jenkins. If the eagle toolchain isn't something that we can throw on
travis-ci could you add it to jenkins?

**Comment by kd0aij** [comment]
@dagar Why does hrt_elapsed_time crash when given an unaligned pointer, when (now - _home.timestamp) runs OK?

**Comment by NaterGator** [comment]
@kd0aij the compiler knows _home.timestamp comes from a packed struct and does assume alignment. When we take a pointer that information is lost and the compiler assumes default alignment.

**Comment by kd0aij** [comment]
Would have saved some time if the compiler warned about such things...

**Comment by NaterGator** [comment]
It is a longstanding issue. Another problem is with [these](https://github.com/PX4/Firmware/pull/4366/commits/936cae132e66ce888a18ea7bd92bcfc982dcf34a#diff-726b3b1c11d1761efd0d9ea754b74a4dR109) size lookups; they're really used to order fields in the struct. First off, they mix and match convention of bytes vs bits (char is 1, int8 is 8... what?) second they do not really account for alignment concerns. Only types whose sizeof(type) % 4 == 0 should be promoted first in the struct, everything else should follow. Adding to that, there is no packing applied to the composite structs that we define there, so we can't necessarily assume they are packed unless we explicitly pack them.

The most insidious issue is the timestamp on _home _should_ be word aligned since it comes first in the struct. Even in the case of packed struct members I am surprised to see GCC placing members without word alignment. That just does not make sense to me.

**Comment by NaterGator** [comment]
I would very very strongly argue that we should not try to screw around with the compiler's default alignment requirements (optimally ordering members is another story and totally fair game). 

Get rid of all aligned/packed attributes and use the python codegen to provide packed serial/deserialization operators for user buffers similar to how mavlink does it. The next gen logger can use those operators. Everything else is going to be a headache game of whack-a-mole.

**Comment by dagar** [comment]
That's the plan - https://github.com/dagar/Firmware/blob/newLogger2/msg/templates/uorb/msg.cpp.template#L128

Almost working...

**Comment by julianoes** [comment]
@kd0aij @dagar: sorry, I wrote that down wrongly. It is `baro_timestamp`, not `baro_failure`.

**Comment by LorenzMeier** [comment]
@dagar Ping me when this is vetted. Thanks for hanging in there!

**Comment by dagar** [comment]
The new logger and uorb generation are more intertwined than I originally thought, so I'm going to keep them together in the same PR again. https://github.com/PX4/Firmware/pull/4365
